Values, variables, pointers, and references (#2006)
carbon-language/carbon-lang: docs/design/README.md

types when they are in a type position, like after a `:` in a variable
declaration or the return type after a `->` in a function declaration. Any
expression in a type position must be

> [a constants or symbolic value](#value-phases) so the compiler can resolve

Reviewer: ```suggestion
[a constant or symbolic value](#value-phases) so the compiler can resolve
```

> -   Proposal
>     [#199: String literals](https://github.com/carbon-language/carbon-lang/pull/199)

## Values, objects, and expressions

Carbon has both abstract _values_ and concrete _objects_. Carbon _values_ are
things like `42`, `true`, and `i32` (a type value). Carbon _objects_ have
_storage_ where values can be read and written. Storage also allows taking the
address of an object in memory in Carbon.

> References:
>
> -   [Values, variables, and pointers](values.md)
> -   Proposal
>     [#2006: Values, variables, pointers, and references](https://github.com/carbon-language/carbon-lang/pull/2006)

### Expression categories

> Carbon expressions produce values or reference objects. Every expression has a

Reviewer: I think I'm supposed to parse this as "(produce values) or (reference objects)", not "produce ((values) or (reference objects))", right? If so, I suggest rephrasing to avoid the ambiguity, e.g.

```suggestion
A Carbon expression produces a value or references an object. Every expression has a
```

> Carbon expressions produce values or reference objects. Every expression has a

Reviewer: Might it make sense to list "initialize objects" as a third thing expressions can do? Otherwise, the structure of this paragraph is kind of confusing, because "produce values" and "reference objects" seem to correspond very closely to the value expression and reference expression categories, which makes it surprising that there's also a third category.

Author: Yeah, good suggestion, thanks!

[category](<https://en.wikipedia.org/wiki/Value_(computer_science)#lrvalue>),
similar to [C++](https://en.cppreference.com/w/cpp/language/value_category):

-   [_Value expressions_](values.md#value-expressions) produce abstract,

>     read-only _values_ that cannot be modified or have its address taken.

Reviewer 2: ```suggestion
    read-only _values_ that cannot be modified or have their address taken.
```

-   [_Reference expressions_](values.md#reference-expressions) refer to
    _objects_ with _storage_ where a value may be read or written and the
    object's address can be taken.
-   [_Initializing expressions_](values.md#initializing-expressions) which
    require storage to be provided implicitly when evaluating the expression.
    The expression then initializes an object in that storage. These are used to
    model function returns, which can construct the returned value directly in
    the caller's storage.

Expressions in one category can be converted to any other category when needed.
The primitive conversion steps used are:

> -   A _read_ converts a reference expression into a value expression.

Reviewer: ```suggestion
-   _Value binding_ converts a reference expression into a value expression.
```

the only pointer [operations](#expressions) are:

> -   Dereference: given a pointer `p`, `*p` gives the value `p` points to as an

Reviewer 2: ```suggestion
-   Dereference: given a pointer `p`, `*p` gives the value `p` points to as a
```

    [reference expression](#expression-categories). `p->m` is syntactic sugar
    for `(*p).m`.

> -   Address-of: given an [reference expression](#expression-categories) `x`,

Reviewer 2: ```suggestion
-   Address-of: given a [reference expression](#expression-categories) `x`,
```

Binding patterns default to _`let` bindings_. The `var` keyword is used to make
it a _`var` binding_.

-   The result of a `let` binding is the name is bound to a

>     [value expression](#expression-categories). This means the value cannot be

Reviewer: ```suggestion
-   A `let` binding binds a name to a value, so the name can be used as a
    [value expression](#expression-categories). This means the value cannot be
```

#### `const`

For every type `MyClass`, there is the type `const MyClass` such that:

-   The data representation is the same, so a `MyClass*` value may be implicitly
    converted to a `(const MyClass)*`.
-   A `const MyClass` [reference expression](#expression-categories) may
    automatically convert to a `MyClass` value expression, the same way that a
    `MyClass` reference expression can.
-   If member `x` of `MyClass` has type `T`, then member `x` of `const MyClass`
    has type `const T`.

> -   The API of a `const MyClass` is a subset of `MyClass`, excluding all methods

Reviewer 3: I think this should be changed? I think you say elsewhere that `const MyClass` has the same API as `MyClass`, but the methods taking `[addr self: Self*]` are just not callable (type mismatch since you can't cast `const Self*` to `Self*`).

Author: Tried to address this, let me know if the wording.
